server/journeys/referral/contributory-factors/schemas.ts

```diff
@@ -1,30 +1,43 @@
 import z from 'zod'
 import { Request } from 'express'
+import { AsyncReturnType } from 'type-fest'
 import CsipApiService from '../../../services/csipApi/csipApiService'
+import { ContributoryFactor } from '../../../@types/express'
+import { createSchema } from '../../../middleware/validationMiddleware'
 
 const ERROR_MESSAGE = 'Select the contributory factors'
 
-export const schemaFactory =
-  (csipApiService: CsipApiService) =>
-  async (req: Request): Promise<z.ZodTypeAny> => {
-    const refDataMap = new Map(
-      (await csipApiService.getReferenceData(req, 'contributory-factor-type')).map(itm => [itm.code, itm]),
-    )
-    return z.object({
-      contributoryFactors: z.union(
-        [
-          z
-            .string()
-            .refine(val => refDataMap.has(val), ERROR_MESSAGE)
-            .transform(val => [{ factorType: refDataMap.get(val) }]),
-          z
-            .array(z.string())
-            .refine(val => val.every(element => refDataMap.has(element)), ERROR_MESSAGE)
-            .transform(val => val.map(element => ({ factorType: refDataMap.get(element) }))),
-        ],
-        { message: ERROR_MESSAGE },
-      ),
-    })
-  }
+export const schemaFactory = (csipApiService: CsipApiService) => async (req: Request) => {
+  const refDataMap = new Map(
+    (await csipApiService.getReferenceData(req, 'contributory-factor-type')).map(itm => [itm.code, itm]),
+  )
+  return createSchema({
+    contributoryFactors: z.union(
+      [
+        z.string().transform<ContributoryFactor[]>((val, ctx) => {
+          if (!refDataMap.has(val)) {
+            ctx.addIssue({
```

Comment:
Does `transform` do the same as superRefine, whereby if you call `addIssue`, it detects it and throws an issue out?

Reply:
Also, why the change from two separated refine then transform lines, to one transform block?

Reply:
failing refine does _not_ prevent the transform getting called.
which means if i separate them, i can't just put a `!` on transform result.
in order to get the type right, i have to do the validation in the transform block and return a z.NEVER.

and yes, transform can do what superRefine does. the main difference is that transform's return value will be used for the result value.

Reply:
How come this wasn't an issue before - what changed?

Reply:
it wasn't an issue because SchemaType inference did not work as expected. the inferred type turned out to be just `any`.
the hidden issue only surfaced when the inferred schematype is fixed.

Reply:
![image](https://github.com/user-attachments/assets/953620b9-2252-4b0e-a960-737b557c55dd)
it seems to be fine for me?

Reply:
Typechecking definitely turns it up - I went to `involvement/controller.ts` and changed 
`req.journeyData.referral!.incidentInvolvement = req.body.involvementType`
to
`req.journeyData.referral!.incidentInvolvement = req.body.foo`
and then `npm run typecheck` spits out an error

Reply:
ahhh it wasn't `Awaited` failing the infer, it was `Promise<z.ZodTypeAny>` lol
but `Awaited` doesn't nicely with intellij, so it'd be difficult for me to figure out whether the type works correctly:
![image](https://github.com/user-attachments/assets/ba9315e5-727c-43e0-b99b-daaf8e3fd367)

Reply:
with Promise<z.ZodTypeAny>:
![image](https://github.com/user-attachments/assets/34ecd53f-2e88-4189-add5-deda4abdd212)

Reply:
So given that you removed that type off the schema functions anyway, shouldn't this be fine apart from IntelliJ being silly? Maybe worth checking for an update to intellij? A quick check of their website says TS language server support is only supposed to work properly on Ultimate or better editions. Then we wouldn't need this extra library.

Reply:
testing result shows that AsyncReturnType's only advantage over Awaited<ReturnType> is that it helps intellij show the type of SchemaType correctly. it doesn't affect type check.

and typing req with Request<unknown, unknown, SchemaType> can provide a correctly typed req.body even on intellij IDE.

so it's not worth have this extra dependency of type-fest, especially when only an older version work and we can't update it.

Reply:
intellij version 2024.2 (released this week) just fixed the issue with Awaited. so no issue at all reverting back to Awaited lol